### PR TITLE
Fix an issue when barcode scanner doesn't work on Pixel 2

### PR DESCRIPTION
**Pull request by alexshikov**
This fixes an issue when barcodes are not scanned on Pixel 2. #1461
Now we take into account the padding (stride) of YUV image when converting it to the new NV21 format.

I wasn't able to build and run **expo**. This fix tested and applied based on the fork of [cameraview](https://github.com/alexshikov/cameraview/tree/vision-detector) library

### CODE REVIEW

**Comment by esamelson** [comment]
Hi @alexshikov - thanks very much for this PR, and sorry that it seems to have slipped through the cracks! This particular has been somewhat of a mystery to us, so we're excited to see a possible solution 😄 We'll get this reviewed soon.

File: android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java [hunk 1]
```diff
@@ -170,23 +170,53 @@ public void onReady() {
...
+                if (image.getFormat() == ImageFormat.JPEG) {
+                    ByteBuffer buffer = image.getPlanes()[0].getBuffer();
                     byte[] data = new byte[buffer.remaining()];
                     buffer.get(data);
-                    if (image.getFormat() == ImageFormat.JPEG) {
-                        mCallback.onPictureTaken(data);
-                    } else {
-                        mCallback.onFramePreview(data, image.getWidth(), image.getHeight(), mDisplayOrientation);
-                    }
-                    image.close();
```
Comment (esamelson) [inline android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java]: is it safe to remove this?

File: android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java [hunk 2]
```diff
@@ -170,23 +170,53 @@ public void onReady() {
...
+                    mCallback.onFramePreview(data, w + dw, h, mDisplayOrientation);
                 }
             }
         }
 
     };
 
+    // WARNING
+    // this is a simplified version of a conversion, but it's enough for QR code detection
+    // for a proper conversion U-channel should be extracted into byte array result
```
Comment (esamelson) [inline android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java]: can we do a proper conversion and include the U-channel? I'm not aware of any reason not to and it seems quite simple: https://stackoverflow.com/questions/40090681/android-camera2-api-yuv-420-888-to-jpeg

**Review by esamelson — CHANGES_REQUESTED** [review]
Thanks very much for this @alexshikov - this appears to work & the mystery is solved! 🎉 I have a couple of questions about the PR itself -- in addition to the comments below, could we add this fix to the `Camera1` API as well?

Apologies again for the very long turnaround time on this. Moving forward it will not be nearly as long!

File: android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java (hunk 2 above)
Comment (alexshikov) [reply android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java]: Sure.

File: android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java (hunk 1 above)
Comment (alexshikov) [reply android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java]: To be honest, I don't remember why I've removed it. I need to re-test having `close()` reverted.

Overall this code works good (we use it in the production app and there are no more issues with the camera/scanner). So seems `image.close()` doesn't lead to memory leaks or other issues.

**Comment by alexshikov** [comment]
> could we add this fix to the `Camera1` API as well?

I've not tested `Camera1` so I'm not sure if it has the same issue. Is there the same issue?

And if I understand the question correct, it's about if _I_ can apply the fix to `Camera1`, right? :)

**Comment by esamelson** [comment]
@alexshikov yes, the same issue is present in `Camera1` as well. And yes, sorry for being unclear 😅 If you aren't able to apply the fix to `Camera1` we can still merge this (better to have this fix in Camera2 than not at all) but it would be even better if you could fix this across the board 🎉

File: android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java (hunk 1 above)
Comment (alexshikov) [reply android/cameraview/cameraview/src/main/api21/com/google/android/cameraview/Camera2.java]: So I get the next hint from the Android Studio when I return `image.close()`.

```
Redundant close less... (⌘F1) 
Inspection reports unnecessary close of resource in the end of try-with-resources block
```

Seems it's safe to remove it.

**Comment by alexshikov** [comment]
Added a proper YUV to NV21 conversion.

As for the `Camera1` issue:
1. I don't have any problems with that and QR-Codes are scanned just fine using the same demo project.
2. Even if there is an issue, it has another nature as Camera1 returns data as NV21 format with proper width/height. The issue with Camera2 related to the invalid width/height calculation after YUV conversion to NV21.

**Comment by esamelson** [comment]
Looks great -- thank you so much @alexshikov ! Could you remove the "WARNING" comment as it no longer applies, and then we can get this merged 😃

**Comment by alexshikov** [comment]
Sure. Just missed this.

**Review by esamelson — APPROVED** [review]
thanks very much @alexshikov !

**Pull request merged**